{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "y0qq9HYfFJtu"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import numpy.random as rand\n",
        "\n",
        "#Simulates univariate data based on demo mutation rates from code in original TADA paper (https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003671)\n",
        "\n",
        "PLOIDY = 2\n",
        "\n",
        "def gen_univ_data(gamma: float, q: float, N_family: int, N_case: int, N_control: int, mut_1: 'array_like', mut_2: 'array_like', seed: int = 0):\n",
        "    '''\n",
        "    Function for generating synthetic data according to TADA probabilistic model.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    gamma : float\n",
        "        Genotype relative risk for genotype Aa where a is an allele with\n",
        "        a deleterious mutation, and A without\n",
        "\n",
        "    q : float\n",
        "        Probability of mutation\n",
        "\n",
        "    N_family : int\n",
        "        Number of families in sample\n",
        "\n",
        "    N_case : int\n",
        "        Number of cases in case/control data\n",
        "\n",
        "    N_control : int\n",
        "        Number of controls in case/control data\n",
        "\n",
        "    mut_1 : array_like\n",
        "        1D array_like of mutation rates for first mutation class with type 'float'\n",
        "\n",
        "    mut_2 : array_like\n",
        "        1D array_like of mutation rates for second mutation class with type 'float'\n",
        "\n",
        "    seed : int\n",
        "        Random seed, defaults to 0\n",
        "\n",
        "    Returns:\n",
        "        df : DataFrame\n",
        "            Generated Data\n",
        "    '''\n",
        "    #Sets random seed\n",
        "    rand.seed(int(seed))\n",
        "\n",
        "    #Ensures mutation rates for the different classes are saved as numpy arrays\n",
        "    mut_1 = np.array(mut_1)\n",
        "    mut_2 = np.array(mut_2)\n",
        "\n",
        "    #Rates for poisson distributions followed by the de novo mutations for the two categories,\n",
        "    # transmitted/nontransmitted data, and transmission counts for case-control data respectively\n",
        "    X_d1_rates = PLOIDY * mut_1 * gamma * N_family\n",
        "    X_d2_rates = PLOIDY * mut_2 * gamma * N_family\n",
        "    X_t_rates = SAMPLE_SIZE * [q * gamma * N_family]\n",
        "    X_nt_rates = SAMPLE_SIZE * [q * N_family]\n",
        "    X_ctrl_rates = SAMPLE_SIZE * [q * (N_control+N_family)]\n",
        "    X_case_rates = SAMPLE_SIZE * [q * gamma * (N_case+N_family)]\n",
        "\n",
        "    #Stacks poisson rates and samples (cls indicates mutation class)\n",
        "    cls1 = np.column_stack([X_d1_rates,np.array(X_t_rates),np.array(X_nt_rates),np.array(X_case_rates),np.array(X_ctrl_rates)])\n",
        "    cls2 = np.column_stack([X_d2_rates,np.array(X_t_rates),np.array(X_nt_rates),np.array(X_case_rates),np.array(X_ctrl_rates)])\n",
        "    sample_1 = rand.poisson(cls1).astype(int)\n",
        "    sample_2 = rand.poisson(cls2).astype(int)\n",
        "\n",
        "    #Combines simulated data with mutation rates into table\n",
        "    table = np.column_stack((sample_1,sample_2))\n",
        "    df = pd.DataFrame(table)\n",
        "    df.columns = ['dn.cls1', 'trans.cls1', 'ntrans.cls1', 'case.cls1', 'ctrl.cls1', 'dn.cls2', 'trans.cls2', 'ntrans.cls2', 'case.cls2', 'ctrl.cls2']\n",
        "\n",
        "    df.insert(0, 'mut.cls1', mut_1.reshape(-1,1))\n",
        "    df.insert(6, 'mut.cls2', mut_2.reshape(-1,1))\n",
        "    return df\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Tests De novo + case/control\n",
        "#Read mutation data\n",
        "tada_file = \"TADA_demo_counts_de-novo_and_inherited.txt\"\n",
        "tada_data = pd.read_table(tada_file)\n",
        "\n",
        "#Specify the number of families and the number of cases and control samples included in the simulation\n",
        "N_family = 4500\n",
        "N_case = 1000\n",
        "N_ctrl = 3000\n",
        "SAMPLE_SIZE = 5000\n",
        "\n",
        "#Extract mutation rates\n",
        "mut_cls1 = tada_data[\"mut.cls1\"]\n",
        "mut_cls2 = tada_data[\"mut.cls2\"]\n",
        "\n",
        "\n",
        "#Varying gamma and q parameters (where q/2 is the allele frequency of a)\n",
        "gamma_1, gamma_2, gamma_3 = 15, 20, 25\n",
        "q_1, q_2, q_3 = 5*10**-5, 10**-4, 2*10**-4\n",
        "\n",
        "df_1 = gen_univ_data(gamma_3, q_3, N_family, N_case, N_ctrl, mut_cls1, mut_cls2)\n",
        "\n",
        "gene_names = tada_data[\"gene.id\"]\n",
        "df_1.insert(0,\"gene.id\",gene_names)\n",
        "\n",
        "#Exports output to text file\n",
        "df_1.to_csv('testUnivData3.txt', index=False, sep='\\t')"
      ],
      "metadata": {
        "id": "XjjzDgJ1NaXd"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import numpy.random as rand\n",
        "\n",
        "#Simulates de novo univariate data based on demo mutation rates from code in original TADA paper (https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003671)\n",
        "\n",
        "PLOIDY = 2\n",
        "SAMPLE_SIZE = 5000\n",
        "\n",
        "def gen_den_univ_data(gamma: float, q: float, N_family: int, mut_1: 'array_like', mut_2: 'array_like', seed: int = 0):\n",
        "    '''\n",
        "    Function for generating synthetic data according to TADA probabilistic model.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    gamma : float\n",
        "        Genotype relative risk for genotype Aa where a is an allele with\n",
        "        a deleterious mutation, and A without\n",
        "\n",
        "    q : float\n",
        "        Probability of mutation\n",
        "\n",
        "    N_family : int\n",
        "        Number of families in sample\n",
        "\n",
        "    mut_1 : array_like\n",
        "        1D array_like of mutation rates for first mutation class with type 'float'\n",
        "\n",
        "    mut_2 : array_like\n",
        "        1D array_like of mutation rates for second mutation class with type 'float'\n",
        "\n",
        "    seed : int\n",
        "        Random seed, defaults to 0\n",
        "\n",
        "    Returns:\n",
        "        df : DataFrame\n",
        "            Generated Data\n",
        "    '''\n",
        "    #Sets random seed\n",
        "    rand.seed(int(seed))\n",
        "\n",
        "    #Ensures mutation rates for the different classes are saved as numpy arrays\n",
        "    mut_1 = np.array(mut_1)\n",
        "    mut_2 = np.array(mut_2)\n",
        "\n",
        "    #Rates for poisson distributions followed by the de novo mutations for the two categories,\n",
        "    # transmitted/nontransmitted data, and transmission counts for case-control data respectively\n",
        "    X_d1_rates = PLOIDY * mut_1 * gamma * N_family\n",
        "    X_d2_rates = PLOIDY * mut_2 * gamma * N_family\n",
        "    X_t_rates = SAMPLE_SIZE * [q * gamma * N_family]\n",
        "    X_nt_rates = SAMPLE_SIZE * [q * N_family]\n",
        "\n",
        "    #Stacks poisson rates and samples (cls indicates mutation class)\n",
        "    cls1 = np.column_stack([X_d1_rates,np.array(X_t_rates),np.array(X_nt_rates)])\n",
        "    cls2 = np.column_stack([X_d2_rates,np.array(X_t_rates),np.array(X_nt_rates)])\n",
        "    sample_1 = rand.poisson(cls1).astype(int)\n",
        "    sample_2 = rand.poisson(cls2).astype(int)\n",
        "\n",
        "    #Combines simulated data with mutation rates into table\n",
        "    table = np.column_stack((sample_1,sample_2))\n",
        "    df = pd.DataFrame(table)\n",
        "    df.columns = ['dn.cls1', 'trans.cls1', 'ntrans.cls1', 'dn.cls2', 'trans.cls2', 'ntrans.cls2',]\n",
        "\n",
        "    df.insert(0, 'mut.cls1', mut_1.reshape(-1,1))\n",
        "    df.insert(4, 'mut.cls2', mut_2.reshape(-1,1))\n",
        "    return df\n",
        "\n"
      ],
      "metadata": {
        "id": "XrlbwIuiKR-T"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#De novo only testing\n",
        "#Read mutation data\n",
        "tada_file = \"TADA_demo_counts_de-novo_and_inherited.txt\"\n",
        "tada_data = pd.read_table(tada_file)\n",
        "\n",
        "#Specify the number of families and the number of cases and control samples included in the simulation\n",
        "N_family = 4500\n",
        "N_case = 1000\n",
        "N_ctrl = 3000\n",
        "\n",
        "#Extract mutation rates\n",
        "mut_cls1 = tada_data[\"mut.cls1\"]\n",
        "mut_cls2 = tada_data[\"mut.cls2\"]\n",
        "\n",
        "\n",
        "#Varying gamma and q parameters (where q/2 is the allele frequency of a)\n",
        "gamma_1, gamma_2, gamma_3 = 15, 20, 25\n",
        "q_1, q_2, q_3 = 5*10**-5, 10**-4, 2*10**-4\n",
        "#Generates data with gamma_1 and q_1\n",
        "df_1 = gen_den_univ_data(gamma_1, q_1, N_family, mut_cls1, mut_cls2)\n",
        "\n",
        "gene_names = tada_data[\"gene.id\"]\n",
        "df_1.insert(0,\"gene.id\",gene_names)\n",
        "\n",
        "#Exports output to text file\n",
        "df_1.to_csv('testDenUnivData.txt', index=False, sep='\\t')"
      ],
      "metadata": {
        "id": "MMi88nhbKS8d"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#WIP Version\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import numpy.random as rand\n",
        "\n",
        "#Simulates univariate data based on demo mutation rates from code in original TADA paper (https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003671)\n",
        "\n",
        "PLOIDY = 2\n",
        "\n",
        "def gen_univ_data(gamma: float, q: float, N_family: int, N_case: int, N_control: int, mut_1: 'array_like', mut_2: 'array_like', seed: int = 0):\n",
        "    '''\n",
        "    Function for generating synthetic data according to TADA probabilistic model.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    gamma : float\n",
        "        Genotype relative risk for genotype Aa where a is an allele with\n",
        "        a deleterious mutation, and A without\n",
        "\n",
        "    q : float\n",
        "        Probability of mutation\n",
        "\n",
        "    N_family : int\n",
        "        Number of families in sample\n",
        "\n",
        "    N_case : int\n",
        "        Number of cases in case/control data\n",
        "\n",
        "    N_control : int\n",
        "        Number of controls in case/control data\n",
        "\n",
        "    mut_1 : array_like\n",
        "        1D array_like of mutation rates for first mutation class with type 'float'\n",
        "\n",
        "    mut_2 : array_like\n",
        "        1D array_like of mutation rates for second mutation class with type 'float'\n",
        "\n",
        "    seed : int\n",
        "        Random seed, defaults to 0\n",
        "\n",
        "    Returns:\n",
        "        df : DataFrame\n",
        "            Generated Data\n",
        "    '''\n",
        "    #Sets random seed\n",
        "    rand.seed(int(seed))\n",
        "\n",
        "    #Ensures mutation rates for the different classes are saved as numpy arrays\n",
        "    mut_1 = np.array(mut_1)\n",
        "    mut_2 = np.array(mut_2)\n",
        "\n",
        "    #Rates for poisson distributions followed by the de novo mutations for the two categories,\n",
        "    # transmitted/nontransmitted data, and transmission counts for case-control data respectively\n",
        "    gamma2 = 4.7\n",
        "    q2 = 6*10**-4\n",
        "    X_d1_rates = PLOIDY * mut_1 * 20 * N_family\n",
        "    X_d2_rates = PLOIDY * mut_2 * 4.7 * N_family\n",
        "    X_t_rates = SAMPLE_SIZE * [q * 2.3 * N_family]\n",
        "\n",
        "    X_t_rates2 = SAMPLE_SIZE * [q2 * N_family]\n",
        "    X_nt_rates = SAMPLE_SIZE * [q * N_family]\n",
        "    X_ctrl_rates = SAMPLE_SIZE * [q * (N_control)]\n",
        "    X_case_rates = SAMPLE_SIZE * [q * gamma * (N_case)]\n",
        "\n",
        "    X_nt_rates2 = SAMPLE_SIZE * [q2 * N_family]\n",
        "    X_ctrl_rates2 = SAMPLE_SIZE * [q2 * (N_control)]\n",
        "    X_case_rates2 = SAMPLE_SIZE * [q2 * 4.7 * (N_case)]\n",
        "\n",
        "    #Stacks poisson rates and samples (cls indicates mutation class)\n",
        "    cls1 = np.column_stack([X_d1_rates,np.array(X_t_rates),np.array(X_nt_rates),np.array(X_ctrl_rates),np.array(X_case_rates)])\n",
        "    cls2 = np.column_stack([X_d2_rates,np.array(X_t_rates2),np.array(X_nt_rates2),np.array(X_ctrl_rates2),np.array(X_case_rates2)])\n",
        "    sample_1 = rand.poisson(cls1).astype(int)\n",
        "    sample_2 = rand.poisson(cls2).astype(int)\n",
        "\n",
        "    #Combines simulated data with mutation rates into table\n",
        "    table = np.column_stack((sample_1,sample_2))\n",
        "    df = pd.DataFrame(table)\n",
        "    df.columns = ['dn.cls1', 'trans.cls1', 'ntrans.cls1', 'case.cls1', 'ctrl.cls1', 'dn.cls2', 'trans.cls2', 'ntrans.cls2', 'case.cls2', 'ctrl.cls2']\n",
        "\n",
        "    df.insert(0, 'mut.cls1', mut_1.reshape(-1,1))\n",
        "    df.insert(6, 'mut.cls2', mut_2.reshape(-1,1))\n",
        "    return df\n",
        "\n",
        "#Read mutation data\n",
        "tada_file = \"TADA_demo_counts_de-novo_and_inherited.txt\"\n",
        "tada_data = pd.read_table(tada_file)\n",
        "\n",
        "#Specify the number of families and the number of cases and control samples included in the simulation\n",
        "N_family = 4500\n",
        "N_case = 1000\n",
        "N_ctrl = 3000\n",
        "SAMPLE_SIZE = 5000\n",
        "\n",
        "#Extract mutation rates\n",
        "mut_cls1 = tada_data[\"mut.cls1\"]\n",
        "mut_cls2 = tada_data[\"mut.cls2\"]\n",
        "\n",
        "\n",
        "#Varying gamma and q parameters (where q/2 is the allele frequency of a)\n",
        "gamma_1, gamma_2, gamma_3 = 15, 20, 25\n",
        "q_1, q_2, q_3 = 5*10**-5, 10**-4, 2*10**-4\n",
        "\n",
        "#Generates data with gamma_1 and q_1\n",
        "df_1 = gen_univ_data(gamma_2, q_2, N_family, N_case, N_ctrl, mut_cls1, mut_cls2)\n",
        "\n",
        "gene_names = tada_data[\"gene.id\"]\n",
        "df_1.insert(0,\"gene.id\",gene_names)\n",
        "\n",
        "#Exports output to text file\n",
        "df_1.to_csv('testUnivData1.txt', index=False, sep='\\t')\n"
      ],
      "metadata": {
        "id": "hssjZtxVvSAL"
      },
      "execution_count": 6,
      "outputs": []
    }
  ]
}