{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "from typing import Any, Tuple, Optional\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from itertools import product\n",
        "import torch\n",
        "import collections\n",
        "from scipy.stats import nbinom, gamma, poisson, hypergeom\n",
        "import scipy.integrate as integrate\n",
        "import os\n",
        "import copy\n",
        "import multiprocessing\n",
        "import uuid\n",
        "from math import exp, log\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "from tqdm import tqdm\n",
        "\n",
        "# Based on original TADA software (He et al. 2013) and rewrite (Klei 2015)\n",
        "\n",
        "# Set seed for reproducibility\n",
        "np.random.seed(100)\n",
        "\n",
        "def TADA(tada_counts, sample_counts, mu, hyperpar, denovo_only, mu_frac=1, pi_gene=1):\n",
        "    # Genome-wide application of TADA for K classes of variants\n",
        "    # tada_counts: Dictionary of K data frames, each consisting of vectors for counts for denovo, case, and control mutation counts\n",
        "    # sample_counts: Dictionary of K data frames, each consisting of a vector with three entries of total sample counts,\n",
        "    #                 one for denovo (# trios), cases (# cases + # trios), and controls (# controls + # trios)\n",
        "    # mu: Data frame with K vectors of mutation rates, one for each mutation category\n",
        "    # mu_frac: Data frame with the fraction to use for each mutation category K\n",
        "    # hyperpar: Dictionary of K data frames, each consisting of entries for gamma.mean.dn, beta.dn, gamma.mean.CC, beta.CC, rho1, nu1, rho0, nu0\n",
        "    # denovo_only: Data frame with K Boolean variables indicating whether only denovo counts should be used (T) or whether both denovo and case-control counts be used (F)\n",
        "    # pi_gene: Data frame with K vectors of estimated fractions of causal variants, one for each class of variants. These fractions will be used to set gene-specific RR (case-control)\n",
        "    # Output: Data frame with BF for each of the K classes of variants as well as BF.total. One entry for each of the genes.\n",
        "\n",
        "    # Make sure every list and dataframe has the same elements\n",
        "    print(\"CALCULATION OF TADA TEST STATISTICS\")\n",
        "    print(\"checking the input for consistent variable names\")\n",
        "    mutation_types = list(tada_counts.keys())\n",
        "    n_mutation = len(mutation_types)\n",
        "    n_samples = len(list(tada_counts.values())[0])\n",
        "\n",
        "    # Make sure mu_frac and pi_gene are data frames\n",
        "    if not isinstance(mu_frac, pd.DataFrame):\n",
        "        mu_frac = pd.DataFrame(np.reshape([[mu_frac] * n_mutation],(1,n_mutation)), columns=mutation_types)\n",
        "\n",
        "    if not isinstance(pi_gene, pd.DataFrame):\n",
        "        pi_gene = pd.DataFrame(np.reshape([[pi_gene] * n_mutation * n_samples],(n_samples,n_mutation,)), columns=mutation_types)\n",
        "\n",
        "    if (\n",
        "        sum([mutation_type in mu for mutation_type in mutation_types]) != n_mutation\n",
        "        or sum([mutation_type in mu_frac for mutation_type in mutation_types]) != n_mutation\n",
        "        or sum([mutation_type in hyperpar for mutation_type in mutation_types]) != n_mutation\n",
        "        or sum([mutation_type in pi_gene for mutation_type in mutation_types]) != n_mutation\n",
        "        or sum([mutation_type in denovo_only for mutation_type in mutation_types]) != n_mutation\n",
        "        or sum([mutation_type in sample_counts for mutation_type in mutation_types]) != n_mutation\n",
        "    ):\n",
        "        return \"mismatch in names for the different variables\"\n",
        "    names_trans_categories = list(sample_counts.keys())\n",
        "    names_N = ['dn', 'ca', 'cn']\n",
        "\n",
        "    for mutation in mutation_types:\n",
        "        if sum(names_N[i] in tada_counts[mutation] for i in range(3)) != 3:\n",
        "            return f\"columns of {mutation} do not match the required 'dn' 'ca' 'cn'\"\n",
        "\n",
        "    # Find the number of genes and the number of different kinds of mutations\n",
        "    n_gene = len(mu)  # was m\n",
        "    n_mutation = len(mutation_types)  # was K\n",
        "\n",
        "    BF = pd.DataFrame()\n",
        "    for mutation in mutation_types:\n",
        "        print(f\"working on :: {mutation}\")\n",
        "        BF_mut = np.array([])\n",
        "        for i in range(len(tada_counts[mutation])):\n",
        "            test_BF = calculate_BF(i,tada_counts[mutation], sample_counts[mutation], mu[mutation], mu_frac[mutation], hyperpar[mutation], denovo_only[mutation].item(), pi_gene[mutation])\n",
        "            BF_mut = np.append(BF_mut,test_BF)\n",
        "\n",
        "        BF = pd.concat([BF, pd.DataFrame(BF_mut)], axis=1)\n",
        "    BF.columns = mutation_types\n",
        "    BF.index = tada_counts[mutation].index\n",
        "\n",
        "    # Calculate the overall BF\n",
        "    BF_total = np.exp(np.log(BF).sum(axis=1))\n",
        "    BF['BF.total'] = BF_total\n",
        "    return BF\n",
        "\n",
        "def TADAnull(tada_counts, sample_counts, mu, hyperpar, denovo_only, mu_frac=1, n_rep=100, dn_max=20, ca_max=200, cn_max=200, max_gap=50):\n",
        "    # Genome-wide application of TADA for K classes of variants\n",
        "    # This function determines the distribution of the null hypothesis test statistics which in turn can be used to determine approximate p-values\n",
        "    # tada.counts: list of K data frames in which each dataframe consists of vectors for counts for denovo, case and control mutation counts\n",
        "    # sample.counts: list of K data frames in which each dataframe consists of a vector with three entries of total sample counts,\n",
        "    #                 one for denovo (# trios), cases (# cases + # trios) and controls (# controls + # trios)\n",
        "    # mu: data frame with K vectors of mutation rates. One for each mutation category.\n",
        "    # mu.frac: data frame with the fraction to use for each mutation category K\n",
        "    # hyperpar: list of K data frames. Each data frame consists of entries for gamma.mean.dn, beta.dn, gamma.mean.CC, beta.CC, rho1, nu1, rho0, nu0\n",
        "    # denovo.only: data frame with K Boolean variables of indicating whether only denovo counts should be used (T) or whether both denovo and case-control\n",
        "    # counts be used (F).\n",
        "    # n_rep: number of repetitions to use recommended to be at least 100. For smaller numbers of genes n_rep should be increased.\n",
        "    # dn.max: number of denovo events for which the BF is pre-computed and stored in a table. This speeds up the simulation process. The function will use dn.max or\n",
        "    #         the maximum number of denovo events for a gene, whichever is smaller.\n",
        "    # ca.max and cn.max: as dn.max this is used to pre-compute a table of common case-control count events to speed up processing. The larger these numbers, the longer the\n",
        "    #                     the longer the pre computation step takes. For very large values of ca.max and cn.max an intergration error might occur.\n",
        "    # max.gap: this is used internally to control the size of the pre-compution of the case-control BF matrix. It represents the gap between two genes in\n",
        "    #           case-control count when ordered from smallest to largest count. This essentially identifies outlying values that hardly ever happen and do not have to be\n",
        "    #           pre-computed.\n",
        "    # Output: dataframe with BFnull for each of the K classes of variants as well as BFnull.total. One entry for each of the genes times n_rep.\n",
        "\n",
        "    print(\"CALCULATION OF TADA TEST STATISTICS UNDER THE NULL HYPOTHESIS\")\n",
        "    mutation_types = list(tada_counts.keys())\n",
        "    n_mutation = len(mutation_types)\n",
        "    n_samples = len(list(tada_counts.values())[0])\n",
        "\n",
        "    # Make sure mu_frac and pi_gene are data frames\n",
        "    if not isinstance(mu_frac, pd.DataFrame):\n",
        "        mu_frac = pd.DataFrame(np.reshape([[mu_frac] * n_mutation],(1,n_mutation)), columns=mutation_types)\n",
        "\n",
        "    #Pre-compute the bayes factors for the denovo data\n",
        "    table_BF_dn = {}\n",
        "    for mutation in mutation_types:\n",
        "        print(f\"working on creating DN table for :: {mutation}\")\n",
        "        x = np.arange(dn_max + 1)\n",
        "        counts = tada_counts[mutation]['dn']\n",
        "        param = hyperpar[mutation]\n",
        "        n = sample_counts[mutation]['dn']\n",
        "        BF = np.column_stack([bayes_factor_dn(x[i], n_dn=n.item(), mu=mu[mutation] * mu_frac[mutation].item(), gamma_dn=param[\"gamma.mean.dn\"].item(), beta_dn=param[\"beta.dn\"].item()) for i in range(len(x))])\n",
        "\n",
        "        table_BF_dn[mutation] = pd.DataFrame(BF, columns=[f\"X{value}\" for value in x])\n",
        "        tada_counts[mutation][\"Ncc\"] = tada_counts[mutation][[\"ca\", \"cn\"]].sum(axis=1)\n",
        "\n",
        "\n",
        "    table_BF_cc = {}\n",
        "    for mutation in mutation_types:\n",
        "        if not denovo_only[mutation].item():\n",
        "            print(f\"working on creating CC table for :: {mutation}\")\n",
        "            tada_counts[mutation][\"Ncc\"] = tada_counts[mutation][[\"ca\", \"cn\"]].sum(axis=1)\n",
        "            Ncc = sorted(tada_counts[mutation][\"Ncc\"])\n",
        "            Ncc_gaps = np.diff(Ncc)\n",
        "            i_gap = Ncc_gaps[np.where(Ncc_gaps > max_gap)[0][0]] if len(np.where(Ncc_gaps > max_gap)[0]) > 0 else max(ca_max,cn_max)\n",
        "            n_ca = min(ca_max, i_gap)\n",
        "\n",
        "            n_ca = ca_max\n",
        "            n_cn = min(cn_max, i_gap)\n",
        "            n_cn = cn_max\n",
        "            if(mutation == 'cls2'):\n",
        "                n_ca = 162\n",
        "                n_cn = 162\n",
        "            x = pd.DataFrame(list(product(range(n_ca + 1), range(n_cn + 1))), columns=[\"ca\", \"cn\"])\n",
        "            param = hyperpar[mutation]\n",
        "            n = sample_counts[mutation][[\"ca\", \"cn\"]]\n",
        "            BF = np.column_stack([bayes_factor_cc(x.loc[i], n_cc=n, gamma_cc=param[\"gamma.mean.CC\"], beta_cc=param[\"beta.CC\"], rho1=param[\"rho1\"], nu1=param[\"nu1\"], rho0=param[\"rho0\"], nu0=param[\"nu0\"]) for i in range(len(x))])\n",
        "            BF = np.reshape(BF,(n_ca+1,n_cn+1))\n",
        "            table_BF_cc[mutation] = pd.DataFrame(BF)\n",
        "\n",
        "    BF = np.array([])\n",
        "    for mutation in mutation_types:\n",
        "        print(\"working on creating null data for ::\", mutation)\n",
        "        BF_col = [permute_gene(i,mu_rate=mu[mutation] * mu_frac[mutation][0], counts=tada_counts[mutation],\n",
        "                                                        n=sample_counts[mutation], n_rep=n_rep, param=hyperpar[mutation],\n",
        "                                                        denovo_only=denovo_only[mutation],\n",
        "                                                        table_cc=table_BF_cc[mutation], table_dn=table_BF_dn[mutation]) for i in range(len(tada_counts[mutation]))]\n",
        "        if(len(BF) != 0):\n",
        "            BF = np.column_stack((BF,np.ravel(BF_col)))\n",
        "        else:\n",
        "            BF = np.append(BF,BF_col)\n",
        "    BF = pd.DataFrame(BF, index = np.arange(len(BF)),columns=mutation_types)\n",
        "    BF_total = np.exp(np.log(BF).sum(axis=1))\n",
        "\n",
        "    return {'BF_null': BF, 'BF_null.total': BF_total}\n",
        "\n",
        "\n",
        "def calculate_BF(i_gene, counts, n, mu, mu_frac, hyperpar, denovo_only, pi_gene):\n",
        "    # Wrapper so that lapply can be used to determine the BF for a gene and a particular mutation variant\n",
        "    # i_gene: gene of interest\n",
        "    # counts: counts for a particular variant, dataframe with vectors for dn, ca, cn\n",
        "    # n: total samples counts, dataframe with entries for dn, ca, cn\n",
        "    # mu: vector with mutation rates for the variant of interest for each gene\n",
        "    # mu_frac: fraction to multiply mu with for the variant of interest\n",
        "    # hyperpar: dataframe with entries for gamma.mean.dn, beta.dn, gamma.mean.CC, beta.CC, rho1, nu1, rho0, nu0 for the\n",
        "    #           variant of interest\n",
        "    # denovo_only: Boolean vector indicating whether only denovo contribution (denovo_only = True) or a combination of\n",
        "    #              denovo and case-control contributions is to be used (denovo_only=False)\n",
        "    # pi_gene: vector with K vectors of estimated fractions of causal variants, one for each class of variants.\n",
        "    #          These fractions will be used to set gene-specific RR (case-control)\n",
        "\n",
        "    if i_gene % 100 == 0 or i_gene == len(counts):\n",
        "        pbar = tqdm(total=len(counts))\n",
        "\n",
        "    ### set the hyper parameters for this gene\n",
        "    hyperpar_gene = hyperpar.copy()\n",
        "    RR_product = hyperpar_gene['gamma.mean.CC'] * hyperpar_gene['beta.CC']\n",
        "    hyperpar_gene['gamma.mean.CC'] = hyperpar_gene['gamma.mean.CC'] * pi_gene[i_gene] + (1 - pi_gene[i_gene])\n",
        "    hyperpar_gene['beta.CC'] = RR_product / hyperpar_gene['gamma.mean.CC']\n",
        "    # Determine the BAYES factor\n",
        "    BF = bayes_factor(x=counts.iloc[i_gene, :], n=n, mu=mu[i_gene] * mu_frac, param=hyperpar_gene, denovo_only=denovo_only)\n",
        "\n",
        "    if i_gene % 100 == 0 or i_gene == len(counts):\n",
        "        pbar.update(i_gene)\n",
        "        pbar.close()\n",
        "\n",
        "    return BF\n",
        "\n",
        "def bayes_factor(x, n, mu, param, denovo_only):\n",
        "    # Bayes factor of the gene combining de novo and case-control\n",
        "    # x: a list of (dn, ca, cn), counts in de novo, cases and controls\n",
        "    # n: a list of (dn, ca, cn), sample sizes\n",
        "    # param: (gamma.mean.dn, beta.dn, gamma.mean.CC, beta.CC, rho1, nu1, rho0, nu0)\n",
        "    # denovo_only: Boolean indicating whether only denovo contribution (True) or a combination of\n",
        "    #              denovo and case-control contributions is to be used (False)\n",
        "    # Prior distribution of RR in de novo: gamma.dist.dn ~ Gamma(gamma.mean.dn * beta.dn, beta.dn)\n",
        "    # Prior distribution of RR in C/C data: gamma.dist.cc ~ Gamma(gamma.mean.CC * beta.CC, beta.CC)\n",
        "    # Prior distribution of q|H1: Gamma(rho1, nu1)\n",
        "    # Prior distribution of q|H0: Gamma(rho0, nu0)\n",
        "\n",
        "    # Contribution of denovo variants in families\n",
        "    BF_dn = bayes_factor_dn(x_dn=x['dn'], n_dn=n['dn'], mu=mu, gamma_dn=param['gamma.mean.dn'], beta_dn=param['beta.dn'])\n",
        "    if not denovo_only:\n",
        "        # Contribution of variants in cases and controls\n",
        "        BF_cc = bayes_factor_cc(x_cc=x[['ca', 'cn']], n_cc=n[['ca', 'cn']], gamma_cc=param['gamma.mean.CC'], beta_cc=param['beta.CC'],\n",
        "                                rho1=param['rho1'], nu1=param['nu1'], rho0=param['rho0'], nu0=param['nu0'])\n",
        "    else:\n",
        "        BF_cc = 1\n",
        "    # Combine the pieces of information\n",
        "    BF = BF_dn * BF_cc\n",
        "    return BF\n",
        "\n",
        "def bayes_factor_dn(x_dn, n_dn, mu, gamma_dn, beta_dn):\n",
        "    # Bayes factor of de novo counts of a gene\n",
        "    # x_dn: the de novo count\n",
        "    # n_dn: the sample size (number of families)\n",
        "    # mu: the mutation rate (of this type of mutational events)\n",
        "    # Prior distribution of RR: gamma ~ Gamma(gamma_dn * beta_dn, beta_dn)\n",
        "    marg_lik0 = poisson.pmf(x_dn, 2 * n_dn * mu)\n",
        "    marg_lik1 = nbinom.pmf(x_dn, gamma_dn * beta_dn, beta_dn / (beta_dn + 2 * n_dn * mu))\n",
        "    BF = marg_lik1 / marg_lik0\n",
        "    return BF\n",
        "\n",
        "\n",
        "def bayes_factor_cc(x_cc, n_cc, gamma_cc, beta_cc, rho1, nu1, rho0, nu0):\n",
        "    # Bayes factor of the case-control data\n",
        "    # BF_cn and BF_ca: contribution from control and case data, respectively\n",
        "    # Input: the count data x_cc, the sample size n_cc and the parameters gamma_cc, beta_cc, rho1 and nu1\n",
        "    # Prior distribution of RR: gamma ~ Gamma(gamma_cc * beta_cc, beta_cc)\n",
        "    # Prior distribution of q|H1: Gamma(rho1, nu1)\n",
        "    # Prior distribution of q|H0: Gamma(rho0, nu0)\n",
        "    marglik0_cc = evidence_null_cc(x_cc, n_cc, rho0, nu0)\n",
        "    marglik1_cc = evidence_alt_cc(x_cc, n_cc, gamma_cc, beta_cc, rho1, nu1)\n",
        "    BF_cn = marglik1_cc[\"cn\"] / marglik0_cc[\"cn\"]\n",
        "    BF_ca = marglik1_cc[\"ca\"] / marglik0_cc[\"ca\"]\n",
        "    BF = BF_cn * BF_ca\n",
        "    return BF\n",
        "\n",
        "def evidence_null_cc(x_cc, n_cc, rho0, nu0):\n",
        "    # model evidence of case-control data: P(x_1,x_0|H_0)\n",
        "    # Input: the count data x_cc, the sample size n_cc, and the rho0 and nu0\n",
        "    # Prior distribution of q|H0: Gamma(rho0, nu0)\n",
        "    marglik0_ctrl_log = np.log(nbinom.pmf(x_cc[\"cn\"], rho0, nu0 / (nu0 + n_cc[\"cn\"])))\n",
        "    marglik0_case_log = np.log(nbinom.pmf(x_cc[\"ca\"], rho0 + x_cc[\"cn\"], (nu0 + n_cc[\"cn\"]) / (nu0 + n_cc[\"cn\"] + n_cc[\"ca\"])))\n",
        "    marglik0_log = marglik0_ctrl_log + marglik0_case_log\n",
        "\n",
        "    return {\"cn\": np.exp(marglik0_ctrl_log), \"ca\": np.exp(marglik0_case_log), \"total\": np.exp(marglik0_log)}\n",
        "\n",
        "def evidence_alt_cc(x_cc, n_cc, gamma_cc, beta_cc, rho1, nu1, q_lower=1e-8, q_upper=0.1, debug=False):\n",
        "    # model evidence of case-control data: P(x_1,x_0|H_1)\n",
        "    # Input: the count data x_cc, the sample size n_cc, and the parameters gamma_cc, beta_cc, rho1, and nu1\n",
        "    # Prior distribution of RR: gamma ~ Gamma(gamma_cc*beta_cc, bet a_cc)\n",
        "    # Prior distribution of q|H1: Gamma(rho1, nu1)\n",
        "\n",
        "    def integrand(u, x_ca, gamma_cc, beta_cc, n_ca, x_cn, rho1, nu1, n_cn):\n",
        "        q = np.exp(u)\n",
        "        return (nbinom.pmf(x_ca, gamma_cc * beta_cc, beta_cc / (beta_cc + n_ca * q)) *\n",
        "            gamma.pdf(q, rho1 + x_cn, scale=1 / (nu1 + n_cn)) *\n",
        "            np.exp(u))\n",
        "\n",
        "    marglik1_ctrl = nbinom.pmf(x_cc[\"cn\"], rho1, nu1 / (nu1 + n_cc[\"cn\"]))\n",
        "    marglik1_case = integrate.quad(integrand, np.log(q_lower), np.log(q_upper),args=(x_cc[\"ca\"].item(), gamma_cc.item(), beta_cc.item(), n_cc[\"ca\"].item(), x_cc[\"cn\"].item(), rho1.item(), nu1.item(), n_cc[\"cn\"].item()))[0]\n",
        "    marglik1 = marglik1_ctrl * marglik1_case\n",
        "    return {\"cn\": marglik1_ctrl, \"ca\": marglik1_case, \"total\": marglik1}\n",
        "\n",
        "def table_BF_dn_wrapper(i_x, x, n_dn, mu, gamma_dn, beta_dn):\n",
        "    # a wrapper function used for generating the denovo table, see bayes.factor.dn for a description of the variables\n",
        "    BF = bayes_factor_dn(x[i_x], n_dn=n_dn, mu=mu, gamma_dn=gamma_dn, beta_dn=beta_dn)\n",
        "    return BF\n",
        "\n",
        "\n",
        "def table_BF_cc_wrapper(i_x, x, n_cc, gamma_cc, beta_cc, rho1, nu1, rho0, nu0):\n",
        "    # a wrapper function used for generating the case-control table, see bayes.factor.cc for a description of the variables\n",
        "    if (i_x % 100 == 0 or i_x == len(x)) and len(x) > 1000:\n",
        "        pb = tqdm(total=len(x))\n",
        "        pb.update(i_x)\n",
        "    BF = bayes_factor_cc(x[i_x], n_cc=n_cc, gamma_cc=gamma_cc, beta_cc=beta_cc, rho1=rho1, nu1=nu1, rho0=rho0, nu0=nu0)\n",
        "    return BF\n",
        "\n",
        "\n",
        "def permute_gene(i_gene, mu_rate, counts, n, n_rep, param, denovo_only, table_cc, table_dn):\n",
        "    # Compute permutation BFs of one gene\n",
        "    # mu_rate: the mutation rate of a gene for the variant of interest\n",
        "    # counts: dn, ca, and co counts for the variant of interest to be permuted, this also has a column for Ncc=ca+cn\n",
        "    # n: sample size, for values for de novo, case, control, and case+control\n",
        "    # n_rep: number of permutation_types\n",
        "    # param: set of hyper parameters for the variant of interest.\n",
        "    # table_cc: table of precomputed BFs for case control events of size max.ca by max.cn for the variant of interest\n",
        "    # table_dn: table of precomputed BFs for denovo events of size number of genes by max.dn for the variant of interest\n",
        "    # Output: vector of n_rep BF generated under the null hypothesis\n",
        "\n",
        "    if i_gene % 100 == 0 or i_gene == counts.shape[0]:\n",
        "        print(f\"Progress: {i_gene}/{counts.shape[0]}\")\n",
        "\n",
        "    # generate permutation data for denovo events\n",
        "    sample_dn = np.random.poisson(2 * n['dn'].item() * mu_rate[i_gene], size=n_rep)\n",
        "    # look up the BF value in the table\n",
        "    BF_dn = table_dn.iloc[i_gene].iloc[sample_dn]\n",
        "    if not denovo_only.item():\n",
        "        # when both denovo and case-control BF are needed\n",
        "        # generate permutation data for case-control events\n",
        "        max_ca = table_cc.shape[0]\n",
        "        max_cn = table_cc.shape[1]\n",
        "        sample_ca = np.zeros(n_rep)\n",
        "        sample_cn = np.zeros(n_rep)\n",
        "        sample_ca = np.random.hypergeometric(counts[\"Ncc\"][i_gene], n[\"ca\"] + n[\"cn\"] - counts[\"Ncc\"][i_gene], n[\"ca\"],n_rep)\n",
        "        sample_cn = counts[\"Ncc\"][i_gene] - sample_ca\n",
        "        # find the generated counts that are outside of the pre-computed table\n",
        "        i_na = np.where((sample_ca + 1 > max_ca) | (sample_cn + 1 > max_cn))[0]\n",
        "\n",
        "        if len(i_na) > 0:\n",
        "            # calculate their BF on a case by case basis\n",
        "            BF_na = np.zeros(len(i_na))\n",
        "            for idx, i in enumerate(i_na):\n",
        "                BF_na[idx] = bayes_factor_cc({\"ca\": sample_ca[i], \"cn\": sample_cn[i]},n_cc=n[[\"ca\", \"cn\"]], gamma_cc=param[\"gamma.mean.CC\"], beta_cc=param[\"beta.CC\"], rho1=param[\"rho1\"], nu1=param[\"nu1\"], rho0=param[\"rho0\"], nu0=param[\"nu0\"])\n",
        "\n",
        "        # set the counts outside the range to missing\n",
        "        sample_ca = np.where(sample_ca > max_ca, np.nan, sample_ca).astype(int)\n",
        "        sample_cn = np.where(sample_cn > max_cn, np.nan, sample_cn).astype(int)\n",
        "        # gather the BF values that can be taken from the pre-computed table\n",
        "        BF_cc = np.array([table_cc.iloc[sample_ca[i]][sample_cn[i]] for i in range(n_rep)])\n",
        "\n",
        "        # replace the missing values with the pre-computed ones\n",
        "        i_na = np.where(np.isnan(BF_cc))[0]\n",
        "        if len(i_na) > 0:\n",
        "            BF_cc[i_na] = BF_na\n",
        "\n",
        "    else:\n",
        "        # if denovo only needed then set BF_dn to 1\n",
        "        BF_cc = np.ones(n_rep)\n",
        "\n",
        "    # determine the total BF from the two components\n",
        "    BF = BF_cc * BF_dn\n",
        "\n",
        "    return BF\n",
        "\n",
        "\n",
        "def Bayesian_FDR(BF, pi0):\n",
        "    # Bayesian FDR control (PMID:19822692, Section2.3)\n",
        "    # BF: a vector of BFs\n",
        "    # pi0: the prior probability that the null model is true\n",
        "    # Return: the q-value of each BF, and the number of findings with q below alpha.\n",
        "\n",
        "    # order the BF in decreasing order, need to retain order to get results back in proper order\n",
        "    i_order = np.argsort(-BF).to_numpy()\n",
        "    BF = BF[i_order]\n",
        "\n",
        "    # convert BFs to PPA (posterior probability of alternative model)\n",
        "    pi = 1 - pi0\n",
        "    q = pi * BF / (1 - pi + pi * BF)  # PPA\n",
        "    q0 = 1 - q  # posterior probability of null model\n",
        "    # the FDR at each PPA cutoff\n",
        "    FDR = np.cumsum(q0) / np.arange(1,len(BF)+1)\n",
        "    # reorder to the original order\n",
        "    FDR = FDR[i_order]\n",
        "    return FDR\n",
        "\n",
        "def bayesFactor_pvalue(BF, BF_null):\n",
        "    # determines the p-value for the BF using permutation_types under the null hypothesis BF_null\n",
        "    # BF: vector with bayes factors based on the data\n",
        "    # BF_null: vector with bayes factors based on permuted data\n",
        "\n",
        "    BF_null = np.sort(BF_null)[::-1]\n",
        "    pval = np.searchsorted(-BF_null, -BF) / len(BF_null)\n",
        "    pval[pval == 0] = 0.5 / len(BF_null)\n",
        "\n",
        "    return pval\n",
        "\n",
        "def denovo_MOM(k, N, mu, C, beta, d=2, S=100, max_kvec=None):\n",
        "    # Estimating relative risk and the number of multiple hits from de novo data\n",
        "    # Input:  k - number of disease genes\n",
        "    #         N - sample size\n",
        "    #         mu - mutation rate for all genes\n",
        "    #         C - observed number of de novo events\n",
        "    #         beta - parameter of the prior distribution of gamma\n",
        "    #         d - number of events to use (1 is 1 or more, 2 is 2 or more)\n",
        "    #         S - number of samples to generate per gene\n",
        "    #         max_kvec - used to generate a time line.\n",
        "    # Output: gamma_mean - the average relative risk,\n",
        "    #         M - the expected number of multi-hit genes\n",
        "\n",
        "    if max_kvec is not None:\n",
        "        if k % 100 == 0 or k == max_kvec:\n",
        "            pb = tqdm(total=max_kvec)\n",
        "            pb.update(k)\n",
        "\n",
        "    m = len(mu)  # number of genes\n",
        "\n",
        "    # enrichment of de novo events\n",
        "    nu = C / (2 * N * np.sum(mu))\n",
        "\n",
        "    # MOM estimator of gamma_mean\n",
        "    gamma_mean = (nu - 1) * m / k + 1\n",
        "\n",
        "    # expected M (choose d = 2)\n",
        "    rs = count_multihit(N, mu, k / m, gamma_mean, beta, d=d, S=S)\n",
        "    M = np.sum(rs['M1']) + np.sum(rs['M0'])\n",
        "\n",
        "    return {'gamma.mean': gamma_mean, 'M': M}\n",
        "\n",
        "\n",
        "def count_multihit(N, mu, pi, gamma_mean, beta, d, S):\n",
        "    # Estimate the number of multihit genes in a genome.\n",
        "    # N: sample size\n",
        "    # mu: mutation rate for all genes\n",
        "    # pi: ratio of number of risk genes and total number of genes\n",
        "    # gamma_mean: the average relative risk\n",
        "    # beta: parameter of the prior distribution of gamma\n",
        "    # d: number of events to use (1 is 1 or more, 2 is 2 or more)\n",
        "    # S: number of samples to generate per gene\n",
        "    # Output: M0 - number of multiple-hit genes for the non-risk genes\n",
        "    #         M1 - number of multiple-hit genes for risk genes\n",
        "\n",
        "    m = len(mu)\n",
        "\n",
        "    # M1: the number of causal genes having d or more de novo mutation_types\n",
        "    p_alt = np.column_stack([multihit_prob(mu_i, N, gamma_mean, beta, d=d, S=S) for mu_i in mu])\n",
        "    M1 = m * pi * np.mean(p_alt, axis=0)\n",
        "\n",
        "    # M0: the number of non-causal genes having d or more de novo mutation_types\n",
        "    p_null = np.column_stack([(1 - poisson.cdf(d_i, 2 * N * mu_i)) for mu_i in mu for d_i in d])\n",
        "    p_null = p_null.reshape(m, len(d))\n",
        "    M0 = m * (1 - pi) * np.mean(p_null, axis=0)\n",
        "\n",
        "    result = pd.DataFrame({'d': d, 'M0': M0, 'M1': M1})\n",
        "    return result\n",
        "\n",
        "def multihit_prob(mu, N, gamma_mean, beta, d, S):\n",
        "    # Prob. of having d or more de novo mutation_types under H1\n",
        "    # Use simulation, but could also use analytic form\n",
        "    # mu: mutation rate for a gene\n",
        "    # N: sample size\n",
        "    # gamma_mean: the average relative risk\n",
        "    # beta: parameter of the prior distribution of gamma\n",
        "    # d: number of events to use (1 is 1 or more, 2 is 2 or more)\n",
        "    # S: number of samples to generate per gene\n",
        "    # Output: p - average probability of having d or more de novo mutation_types\n",
        "\n",
        "    gamma = gamma.rvs(gamma_mean * beta, scale=1 / beta, size=S)\n",
        "    p = 1 - poisson.cdf(d, 2 * N * mu * gamma)\n",
        "    return np.mean(p)\n"
      ],
      "metadata": {
        "id": "x7jLX5Uw5XlX"
      },
      "execution_count": 28,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Read mutation data\n",
        "tada_file = \"TADA_smoke_test_counts.txt\"\n",
        "tada_data = pd.read_table(tada_file)\n",
        "\n",
        "# Specify the number of families and the number of cases and control samples included in the analysis\n",
        "n_family = 4500\n",
        "n_case = 1000\n",
        "n_ctrl = 3000\n",
        "\n",
        "data = {'dn': [n_family], 'ca': [n_case + n_family], 'cn': [n_ctrl + n_family]}\n",
        "n = pd.DataFrame(data)\n",
        "sample_counts = {'cls1': n, 'cls2': n}\n",
        "\n",
        "# Create the mutational data used by TADA\n",
        "cls1_counts = pd.DataFrame({'dn': tada_data['dn.cls1'],\n",
        "                            'ca': tada_data['trans.cls1'] + tada_data['case.cls1'],\n",
        "                            'cn': tada_data['ntrans.cls1'] + tada_data['ctrl.cls1']})\n",
        "cls1_counts.index = tada_data['gene.id']\n",
        "\n",
        "cls2_counts = pd.DataFrame({'dn': tada_data['dn.cls2'],\n",
        "                            'ca': tada_data['trans.cls2'] + tada_data['case.cls2'],\n",
        "                            'cn': tada_data['ntrans.cls2'] + tada_data['ctrl.cls2']})\n",
        "cls2_counts.index = tada_data['gene.id']\n",
        "\n",
        "tada_counts = {'cls1': cls1_counts, 'cls2': cls2_counts}\n",
        "\n",
        "# Set up mutation rates\n",
        "mu = pd.DataFrame({'cls1': tada_data['mut.cls1'], 'cls2': tada_data['mut.cls2']})\n",
        "\n",
        "# Set up denovo only TRUE/FALSE, here we do not want to restrict ourselves to de novo only analyses\n",
        "denovo_only = pd.DataFrame({'cls1': [False], 'cls2': [False]})\n",
        "\n",
        "# Set up parameters\n",
        "cls1_params = pd.DataFrame({'gamma.mean.dn': [20.0],\n",
        "                            'beta.dn': [1],\n",
        "                            'gamma.mean.CC': [2.3],\n",
        "                            'beta.CC': [4.0],\n",
        "                            'rho1': [0.1],\n",
        "                            'nu1': [100],\n",
        "                            'rho0': [0.1],\n",
        "                            'nu0': [100]})\n",
        "\n",
        "cls2_params = pd.DataFrame({'gamma.mean.dn': [4.7],\n",
        "                            'beta.dn': [1],\n",
        "                            'gamma.mean.CC': [1.0],\n",
        "                            'beta.CC': [1000],\n",
        "                            'rho1': [0.15],\n",
        "                            'nu1': [100],\n",
        "                            'rho0': [0.15],\n",
        "                            'nu0': [100]})\n",
        "\n",
        "hyperpar = {'cls1': cls1_params, 'cls2': cls2_params}\n",
        "# Running TADA\n",
        "re_TADA = TADA(tada_counts=tada_counts, sample_counts=sample_counts, mu=mu, hyperpar=hyperpar, denovo_only=denovo_only)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Qxow3VYD2qwG",
        "outputId": "3172641e-44de-467b-f148-984e24016a76"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CALCULATION OF TADA TEST STATISTICS\n",
            "checking the input for consistent variable names\n",
            "working on :: cls1\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "  0%|          | 0/10 [00:00<?, ?it/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "working on :: cls2\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "  0%|          | 0/10 [00:00<?, ?it/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Bayesian FDR control\n",
        "re_TADA['qval'] = Bayesian_FDR(re_TADA['BF.total'], pi0=0.95)\n",
        "\n",
        "# Run permutation to get the null distributions to use for calculating p-values for TADA\n",
        "re_TADA_null = TADAnull(tada_counts=tada_counts, sample_counts=sample_counts, mu=mu, hyperpar=hyperpar, denovo_only=denovo_only, n_rep=100)\n",
        "re_TADA['pval'] = bayesFactor_pvalue(re_TADA['BF.total'], re_TADA_null['BF_null.total'])\n",
        "\n",
        "# Top 10 genes based on BF.total\n",
        "re_TADA = re_TADA.sort_values(by='BF.total', ascending=False).head(10)\n",
        "print(re_TADA)"
      ],
      "metadata": {
        "id": "5eqFDg9P5Z_0",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c5dcf2a1-ae66-49bc-c7b9-1f4dc2047b9b"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CALCULATION OF TADA TEST STATISTICS UNDER THE NULL HYPOTHESIS\n",
            "working on creating DN table for :: cls1\n",
            "working on creating DN table for :: cls2\n",
            "working on creating null data for :: cls1\n",
            "Progress: 0/10\n",
            "working on creating null data for :: cls2\n",
            "Progress: 0/10\n",
            "                  cls1           cls2      BF.total          qval    pval\n",
            "gene.id                                                                  \n",
            "GENE956   2.933410e+05  591202.755311  1.734240e+11  1.095581e-10  0.0005\n",
            "GENE2602  5.819384e+07       2.822945  1.642780e+08  5.788358e-08  0.0005\n",
            "GENE4190  1.606000e+05       3.088818  4.960642e+05  1.280526e-05  0.0005\n",
            "GENE2864  3.175408e+03      20.399743  6.477751e+04  8.291037e-05  0.0005\n",
            "GENE2403  2.513713e+01      20.738377  5.213032e+02  7.099416e-03  0.0005\n",
            "GENE1227  9.443343e+00      22.591615  2.133404e+02  1.954561e-02  0.0005\n",
            "GENE1904  9.510587e+00      18.631253  1.771942e+02  3.058807e-02  0.0005\n",
            "GENE1896  1.613211e+02       0.832099  1.342352e+02  4.226361e-02  0.0005\n",
            "GENE598   1.113122e-01     528.746194  5.885592e+01  6.468327e-02  0.0005\n",
            "GENE2381  1.875653e+00      16.503311  3.095449e+01  9.624957e-02  0.0005\n"
          ]
        }
      ]
    }
  ]
}